akka-http-core/src/main/scala/akka/http/impl/model/parser/CommonActions.scala

```diff
@@ -32,24 +31,27 @@ private[parser] trait CommonActions {
         case custom        ⇒ MediaType.customMultipart(custom, params)
       }
       case mainLower ⇒
-        if (areNoCustomMediaTypesDefined) // try to prevent closure creation for usual case without custom media types
-          getPredefinedOrCreateCustom(mainType, subType, charsetDefined, params)
-        else
-          customMediaTypes(mainLower, subLower) getOrElse
-            getPredefinedOrCreateCustom(mainLower, subLower, charsetDefined, params)
-    }
-  }
 
-  private def getPredefinedOrCreateCustom(mainLower: String, subLower: String, charsetDefined: Boolean,
-                                          params: Map[String, String]): MediaType =
-    MediaTypes.getForKey((mainLower, subLower)) match {
-      case Some(registered) ⇒ if (params.isEmpty) registered else registered.withParams(params)
-      case None ⇒
-        if (charsetDefined)
-          MediaType.customWithOpenCharset(mainLower, subLower, params = params, allowArbitrarySubtypes = true)
-        else
-          MediaType.customBinary(mainLower, subLower, MediaType.Compressible, params = params, allowArbitrarySubtypes = true)
+        // Faster version of MediaType.withParams for the common case of empty params
+        def withParams(mt: MediaType): MediaType = if (params.isEmpty) mt else mt.withParams(params)
+
+        // Try user-defined function to get a MediaType
+        customMediaTypes(mainLower, subLower) match {
+          case Some(customMediaType) ⇒ withParams(customMediaType)
```

Comment:
The main change is here: calling `customMediaTypes.withParams(params)`.

Reply:
So after this change, when a request comes in for which a custom media type is defined for the `mainLower/subLower` part, it is selected but its parameters are replaced by the parameters from the incoming request.

This does mean registering a custom `WithFixedCharset` media type becomes weird: if you'd register `customWithFixedCharset("foo", "bar", UTF8)` and a request with `foo/bar; charset=us-ascii` comes in, you'd be left with a `t: WithFixedCharset` where `t.charset` is still UTF8 but `t.params("charset")` is us-ascii.

On the other hand, since custom media types are identified by their `mainLower/subLower` part so you can't have a 'fixed' and an 'open' version of the same type co-exist anyway, accepting the request with a slightly-'weird' media type might be better than rejecting the request outright.